javaslang/src/main/java/javaslang/collection/Tree.java

```diff
@@ -748,12 +771,12 @@ public boolean isEmpty() {
 
         @Override
         public boolean isLeaf() {
-            return children.isEmpty();
+            return size == 1;
         }
 
         @Override
         public int size() {
-            return 1 + children.foldLeft(0, (acc, child) -> acc + child.length());
+            return size;
```

Comment:
`Lazy` is pretty fast already, but can slightly be optimized by minimizing the volatile reads, like:
```Java
@Override
public T get() {
    T result = value; // volatile field is accessed fewer times this way
    if (result == null) {
        synchronized (this) {
            result = value;
            if (result == null) {
                result = value = supplier.get();
                supplier = null; // free mem
            }
        }
    }
    return result;
}
```
though that will imply that the `Supplier` may not return `null`.
The speed advantages would be something like:
```
10000000 Lazys, 100 iterations (lot of GC)

// Before 
eagerGetDuration: 7755
initDuration: 19958
calculationDuration: 379
lazyGetDuration: 15146

vs.

// After
eagerGetDuration: 7307
initDuration: 28623
calculationDuration: 428
lazyGetDuration: 11356
```

or
```
1000000 Lazys, 1000 iterations (less GC)

Before
eagerGetDuration: 8866
initDuration: 60
calculationDuration: 35
lazyGetDuration: 11731


```